airbyte-integrations/connectors/source-mailchimp/source_mailchimp/client.py

```diff
@@ -63,45 +68,87 @@ def lists(self, state: DefaultDict[str, any]) -> Generator[AirbyteMessage, None,
         date_created = self._get_cursor_or_none(state, stream_name, cursor_field)
         default_params, max_date_created = self._get_default_params_and_cursor(cursor_field, date_created)
 
-        offset = 0
-        done = False
-        while not done:
-            params = dict(default_params, count=self.PAGINATION, offset=offset)
-            lists_response = self._client.lists.all(**params)["lists"]
-            for mc_list in lists_response:
-                list_created_at = parser.isoparse(mc_list[cursor_field])
-                max_date_created = max(max_date_created, list_created_at) if max_date_created else list_created_at
-                yield self._record(stream=self._LISTS, data=mc_list)
-
-            if max_date_created:
-                state[self._LISTS][cursor_field] = self._format_date_as_string(max_date_created)
-                yield self._state(state)
-
-            done = len(lists_response) < self.PAGINATION
-            offset += self.PAGINATION
+        generator = self.query_paginated(
+            state=state,
+            stream_name=stream_name,
+            params=dict(default_params, count=self.PAGINATION, offset=0),
+            query_subject=self._client.lists,
+            response_data_field="lists",
+            cursor_field=cursor_field,
+            max_cursor_field_value=max_date_created,
+        )
+        return generator
 
     def campaigns(self, state: DefaultDict[str, any]) -> Generator[AirbyteMessage, None, None]:
+
+        self.reset_campaign_ids()
         cursor_field = "create_time"
         stream_name = self._CAMPAIGNS
         create_time = self._get_cursor_or_none(state, stream_name, cursor_field)
         default_params, max_create_time = self._get_default_params_and_cursor(cursor_field, create_time)
 
-        offset = 0
+        generator = self.query_paginated(
+            state=state,
+            stream_name=stream_name,
+            params=dict(default_params, count=self.PAGINATION, offset=0),
+            query_subject=self._client.campaigns,
+            response_data_field="campaigns",
+            cursor_field=cursor_field,
+            max_cursor_field_value=max_create_time,
+            store_responce_field={"where": self.campaign_ids, "field": "id"},
+        )
+        return generator
+
+    def email_activity(self, state: DefaultDict[str, any]) -> Generator[AirbyteMessage, None, None]:
+        if not self.campaign_ids:
+            _ = self.campaigns(state=state)
```

Comment:
Unfortunately, you can't rely on campaigns sync here, because if sync in incremental mode we will not get updates for campaigns that was already synced.
Lets say we have campaigns:
camp1
camp2
camp3
We run sync and synced camp1 and camp2, so the next time we sync only camp3, so email_activity don't know anything about camp1 and camp2.
Please follow the advice from Sherif and try to use CDK with nested streams.